### PR TITLE
[#37] Allowed config filename

### LINKED ISSUE #37: Configure screenshot name pattern in behat.yml
Different to #28 (which I read to mean, specifying the output filename in the stepdefinition, eg `Then I save screenshot to my-screenshot.png`).

It would be nice to have control over the screenshot filename generation when the filename is not specified. My use case would be the ability to remove the time-based uniqueness, since then the screenshot filename is based on the feature / step (or failing step), and you can hit reload in the browser to see updated HTML / PNG after a test run. Currently this workflow requires re-opening the newly generated file 😁 

Just a wishlist item ;)

### PR DESCRIPTION
**Pull request by tannguyen04**
## Checklist before requesting a review

closes #37 

- [x] I have formatted the subject to include ticket number as `[#123] Verb in past tense with dot at the end.`
- [x] I have added a link to the issue tracker
- [x] I have provided information in `Changed` section about WHY something was done if this was not a normal implementation
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have run new and existing relevant tests locally with my changes, and they passed
- [x] I have provided screenshots, where applicable

## Changed

1.

### CODE REVIEW

**Comment by tannguyen04** [comment]
Hi @AlexSkrypnyk 
I have few notes

1. With url token replacements, I am using url encode to build the token replacement. I can not find the way to follow your logic with url tokens.
2. Step name token should not lower cased because some case is sensitive. Ex: https://github.com/drevops/behat-screenshot/pull/94/files#diff-88ecc745a3641e37dcc57852c6bde91b4a709b0d10026876da436f918feb687dR68

File: src/DrevOps/BehatScreenshotExtension/Context/ScreenshotContext.php [hunk 1]
```diff
@@ -231,17 +249,343 @@ protected function prepareDir(string $dir): void
...
+        } elseif (empty($filename)) {
+            $filename = $this->filenamePattern;
+        }
+
+        // Make sure {ext} token is on filename.
+        if (!str_ends_with($filename, '.{ext}')) {
+            $filename .= '.{ext}';
+        }
+
+        return $this->replaceToken($filename, ['ext' => $ext]);
```
Comment (AlexSkrypnyk) [inline src/DrevOps/BehatScreenshotExtension/Context/ScreenshotContext.php:273]: @tannguyen04 
Please extract all tones in to `Tokenizer` class.

**Review by AlexSkrypnyk — CHANGES_REQUESTED** [review]
@tannguyen04 
Could you please:
1. Extract all token-related functionality in to `Tokenizer` class.
2. Add unit tests for Tokeniser class.

**Review by AlexSkrypnyk — CHANGES_REQUESTED** [review]
Please see my inline comments. The Tokenizer class implementation is not correct.

File: src/DrevOps/BehatScreenshotExtension/Context/ScreenshotContext.php [hunk 2]
```diff
@@ -272,15 +270,26 @@ protected function makeFileName(string $ext, string $filename = null, bool $fail
...
+            $url = $this->getSession()->getCurrentUrl();
+        } catch (\Exception) {
+            $url = null;
+        }
+
         $data = [
             'ext' => $ext,
-            'step_scope' => $this->beforeStepScope,
-            'session' => $this->getSession(),
+            'step_name' => $step->getText(),
```
Comment (AlexSkrypnyk) [inline src/DrevOps/BehatScreenshotExtension/Context/ScreenshotContext.php:285]: 😗👌

**Comment by xurizaemon** [comment]
Great work @tannguyen04! 🎈

**Pull request merged**